content/blog/2022-04-19-tracing/index.md

```diff
@@ -0,0 +1,142 @@
++++
+title = "Trace-Based Just-in-Time Type Specialization for Dynamic Languages"
+[extra]
+latex = true
+[[extra.authors]]
+name = "Yulun Yao"
++++
+
+## Introduction
+# TL.DR.
+This paper presents *Trace-Monkey*, a JavaScript language implementation based on tracing Just-In-Time (JIT) compilation. It focuses on identifying frequently executed loop traces at run-time and then compile those traces with dynamic type information. Specifically, the author proposes an algorithm to lazily form trace-trees with alternative paths within the loop, and explains how to generate efficient type-specialized code for the identified traces. The experimental result shows that *Trace-Monkey* is 10x faster compared to previous compilation strategies on some benchmarks.
+
+# Problem
+Dynamic languages, such as JavaScript, Python and Ruby are some of the top programming languages in the recent decade. These languages are accessible to developers thanks to their less restrictive syntax and type systems, easy to deploy on a wide range of platforms since they contain less hardware-specific type information. On the other hand, however, compared to traditional, statically typed languages, namely Java, C++, Ocaml, dynamic languages are much more difficult to be compiled. No hardware would ever support non-typed arithmetic operations, so the typing information needs to be resolved eventually. As types of expressions may vary at runtime in dynamic languages, the typing information is not available to compilers, which very often need to utilize it to do optimizations and generate efficient bytecodes. A seemingly simple yet not-so-simple example for compiler is as follows:
+
+``` 
+function add(x, y, z) {
+	for (i = 0; i < x.length(); i++){
+        z[i] = x[i] + y[i];
+    }
+    return z;
+}
+```
+
+This function computes the sum(?) of each corresponding `x`, `y` slots and put the computation result into `z`. With a traditional, statically-typed languages, there are tons of optimizations we could do. For example, if we know `x`, `y`, `z` are all integer arrays, we could compile this program into vadd (vector adds) if our ISA supports it to speedup our computation. But what if `x`, `y`, `z` are doubles? Or even strings which need to be concatenated. 
```

Comment:
I'm not sure about the vectorization thing. At least in JavaScript, `+` on arrays does not do elementwise addition on arrays. It does something way, way weirder:

```
$ node                                                                        ~
Welcome to Node.js v18.1.0.
Type ".help" for more information.
> [1, 2, 3] + [4, 5, 6]
'1,2,34,5,6'
```